beacon-chain/blockchain/process_block.go

```diff
@@ -91,6 +91,30 @@ func (s *Service) postBlockProcess(ctx context.Context, signed interfaces.ReadOn
 	}
 	newBlockHeadElapsedTime.Observe(float64(time.Since(start).Milliseconds()))
 
+	if headRoot == blockRoot {
+		// Updating next slot state cache can happen in the background
+		// except in the epoch boundary in which case we lock to handle
+		// the shuffling and proposer caches updates.
+		// We handle these caches only on canonical
+		// blocks, otherwise this will be handled by lateBlockTasks
+		slot := postState.Slot()
+		if slots.IsEpochEnd(slot) {
+			if err := transition.UpdateNextSlotCache(ctx, blockRoot[:], postState); err != nil {
+				return errors.Wrap(err, "could not update next slot state cache")
+			}
+			if err := s.handleEpochBoundary(ctx, slot, postState, blockRoot[:]); err != nil {
+				return errors.Wrap(err, "could not handle epoch boundary")
+			}
+		} else {
+			go func() {
```

Comment:
Should this NOT be in the background, so that `ProcessSlotsUsingNextSlotCache` in `getPayloadAttribute` can use it? Otherwise, we might be doing it twice.

Reply:
Yeah but it's a little tricky, since I am moving to a system in which some of these updates are only done if we are proposing (and not reorging) then it's better to leave to a separate PR if you don't mind.